README.md
#thelightingcontrollerclient

A node.js client library for communicating with The Lighting Controller software products (ShowXpress, QuickDmx, and SweetLight).  This library is not created, supported, or endorsed by Chauvet, Showtec, SweetLight, or The Lighting Controller.

This was created using The Lighting Controller's publicly documented [Protocol Definition](http://download.thelightingcontroller.com/software/External_Application/Protocol.pdf).


1. [Install](#install)
2. [Example](#example)
3. [Instantiation](#instantiation)
4. [Methods](#methods)
 - [client.connect](#clientconnect)
 - [client.autoBpmOn](#clientautobpmon)
 - [client.autoBpmOff](#clientautobpmoff)
 - [client.beat](#clientbeat)
 - [client.bpm](#clientbpmnumber-beatsperminute)
 - [client.buttonList](#clientbuttonlist)
 - [client.buttonPress](#clientbuttonpressstring-buttonname)
 - [client.buttonRelease](#clientbuttonreleasestring-buttonname)
 - [client.buttonToggle](#clientbuttontogglestring-buttonname)
 - [client.cue](#clientcuestring-cuename)
 - [client.freeze](#clientfreeze)
 - [client.unfreeze](#clientunfreeze)
 - [client.faderChange](#clientfaderchangenumber-faderindex-number-fadervalue)
5. [Events](#events)
 - [connected](#connected)
 - [disconnected](#disconnected)
 - [error](#error)
 - [beatOn](#beaton)
 - [beatOff](#beatoff)
 - [bpm](#bpm)
 - [buttonList](#buttonlist)
 - [buttonPress](#buttonpress)
 - [buttonRelease](#buttonRelease)
 - [faderChange](#faderchange)
 - [interfaceChange](#interfacechange)
 - [unknownEvent](#unknownevent)
6. [Development](#development)
7. [Known Live Bugs](#bugs-in-the-live-software)
8. [Unknown Features](#unknown-features)


# Install

``$ npm install thelightingcontrollerclient``


---


# Example

```js
//import the client
const LightingController = require('thelightingcontrollerclient');

//create a new instance of the client
const client = new LightingController({
	password: '1234'
});

//subscribe to certain events, for a full list of events, see below
client.on('connected', () => {
	client.buttonList(); //get the button list
});

//print to console when we get the buttons
client.on('buttonList', (buttons) => {
	console.log(JSON.stringify(buttons, null, 2));
});

client.connect();
```

The client is an instance of [EventEmitter2](https://github.com/asyncly/EventEmitter2). You can use onAny to subscribe to all events and click around in Live to see what is sent to the client.  This also allows for usage of methods like .once, .many, .removeAllListeners, and all other EventEmitter2 methods.  See the EventEmitter2 docs for a list of methods for interacting with the EventEmitter.

```js
client.onAny((event, value) => {
	console.log(event, value);
});
```

A full client example can be found at [tidwell/thelightingcontrollerclient-example](https://github.com/Tidwell/thelightingcontrollerclient-example).


---


# Instantiation

When creating an instance of the client, you should pass a configuration object.  The following defaults have been configured, you should pass any properties that you need to overwrite.

```js
const LightingController = require('thelightingcontrollerclient');
const client = new LightingController({
	port: 7348,
	ip: '127.0.0.1',
	extApp: 'thelightingcontrollerclient',
	password: ''
});
```


---


# Methods

###client.connect()
Establishes a connection to the Live application.  On success, will fire a 'connected' event

```js
 	client.connect();
```


###client.autoBpmOn()
Enables Automatic BPM detection.

```js
 	client.autoBpmOn();
```


###client.autoBpmOff()
Disables Automatic BPM detection.

```js
 	client.autoBpmOff();
```


###client.beat()
Increments the beat in the "auto BPM" section of Live.
Send after receiving 'beatOn' event from Live and stop sending after receiving a 'beatOff' event.

```js
 	client.beat();
```


###client.bpm(Number beatsPerMinute)
Assigns the new BPM value in the "manual BPM" section of Live.
Send at each song start or when recieving a 'bpm' event from Live.

```js
 	client.bpm(120);
```


###client.buttonList()
Send to ask Live for the button and master faders list.  On success, will fire a 'buttonList' event

```js
 	client.buttonList();
```


###client.buttonPress(String buttonName)
Send to press the button with the name {buttonName} in Live.

```js
 	client.buttonPress('button one');
```


###client.buttonRelease(String buttonName)
Send to release the button with the name {buttonName} in Live.

```js
 	client.buttonRelease('button one');
```


###client.buttonToggle(String buttonName)
Send to toggle the button with the name {buttonName} in Live.  This is an alias of .cue()

```js
 	client.buttonToggle('button one');
```


###client.cue(String cueName)
Send to toggle the button with the name {buttonName} in Live.  This is an alias of .buttonToggle()

```js
 	client.cue('button one');
```


###client.freeze()
Send to Freeze the Live board

```js
 	client.freeze();
```


###client.unfreeze()
Send to Un-Freeze the Live board

```js
 	client.unfreeze();
```


###client.faderChange(Number faderIndex, Number faderValue)
Send to change the position of the {faderIndex} fader in Live to {faderValue}.  {faderValue} is a percentage between -100 and 100.  *Note* The Live App expects The fader's index, NOT the name.

```js
 	client.faderChange(1, 50);
```


---


# Events

###connected
The client has connected and successfully authenticated with Live.
 
```js
 	client.on('connected', () => {
	});
```


###disconnected
The client has disconnected from Live.

```js
	client.on('disconnected', () => {
	});
```


###error
The client has encoutered an error.

All errorObjects will have a `type`, one of:
- **BAD PASSWORD** - Response from Live that the password for Live was invalid
- **SOCKET** - The client was unable to establish a connection to Live
- **BUTTON LIST XML PARSE FAILED** - The client recieved an xml response from Live that it was unable to parse
- **UNKNOWN ERROR** - The client threw an error but was unable to determine the type.  Check the other properties of the error object to determine what went wrong.
- **CLIENT ERROR** - Data passed to a method was invalid and was not sent to Live.  The ``error`` proprty will contain a human-readable definition of the error.

errorObjects may also contain an `error` property. This is the original error object thrown by whatever encountered the error.  Not all errors have this property

errorObjects may also contain a `data` property. This is the data that was being processed when the error occured.  Not all errors have this property.

```js
	client.on('error', (errorObject) => {
		//example errorObject for calling client.bpm(-100);
		{
			type: 'CLIENT ERROR',
			error: 'invalid bpm value',
			data: -100
		}
	
	});
```


###beatOn
The client recieved a signal that it can start sending real time beats for live to use in BPM calculations. Use .beat() to respond.  *Note* The AutoBPM feature of the Live software only works on a PC.

```js
	client.on('beatOn', () => {
	});
```


###beatOff
The client recieved a signal that it should stop sending real time beats.  *Note* The AutoBPM feature of the Live software only works on a PC.

```js
 	client.on('beatOff', () => {
	});
```


###bpm
The client recieved a request for the current BPM.  Use .bpm(Number) to respond.

```js
	client.on('bpm', () => {
	});
```


###buttonList
The client recieved a response to calling .buttonList() containing the Live button and master faders list.

```js
	client.on('buttonList', (buttonListObject) => {
		//example buttonListObject:

		{
		   	pages: [
		        {
		            name: String
		            columns: Number
		            columnButtons: {
						1: Number,
						2: Number,
						.
						.
						.
		            },
		            buttons [
		            	{
			            	name: String,
			            	index: Number,
			            	flash: Boolean,
			            	pressed: Boolean,
			            	line: Number,
			            	column: Number,
			            	color: String (hex ex #000000)
			            },
			            .
			            .
			            .
			        ]
		        },
		        .
		        .
		        .
		    ],
			faders: [
				{
		            name: String,
		            value: Boolean
		        },
		        .
		        .
		        .
		    ]
		}
	});
```


###buttonPress
A button was pressed

```js
 	client.on('buttonPress', (buttonName) => {
 		//String buttonName The name of the button that was pressed
	});
```


###buttonRelease
A button was released

```js
 	client.on('buttonPress', (buttonName) => {
		//String buttonName The name of the button that was released
	});
```


###faderChange
A fader was moved in value. *Note* The Live App sends The fader's index, NOT the name.

```js
 	client.on('faderChange', (faderObject) => {
		//Example faderObject:
		{
			index: Number,
			value: Number
		}
	});
```
		

###interfaceChange
The interface has changed in the Live software - generally a hint to call .buttonList()

```js
 	client.on('interfaceChange', () => {
	});
```


###unknownEvent
The client encountered a socketMessage it was unable to parse.  Could potentially occur if Live is upated and the client library has yet to be updated to support new events.  Allows for parsing the message manually.

```js
	client.on('unknownEvent', (socketMessage) => {
		//the raw socket message.  See the Protocol Definition for parsing information.
	});
```


---


# Development

[![Build Status](https://travis-ci.org/Tidwell/thelightingcontrollerclient.svg?branch=master)](https://travis-ci.org/Tidwell/thelightingcontrollerclient)

Clone the repo

```bash
	$ git clone https://github.com/Tidwell/thelightingcontrollerclient.git
```

Run tests

```bash
	$ npm test
```

Run coverage report

```bash
	$ npm run-script coverage
```

Mock Server

If you need to see how the offical [Live Mobile Apps](http://www.thelightingcontroller.com/viewtopic.php?f=47&t=836) sends commands to Live, you can run this script as a mock Live instance.  It will simply dump all commands sent to it into console.

```bash
	$ npm run-script mock-server
```

Echo Example

If you need to just dump commands as they are sent from Live, you can run this script.

```bash
	$ npm run-script example
```

---


# Bugs in the Live Software

A bug report has been opened for these issues with TheLightingController, but until they are fixed in the software, you may encounter these issues:

- Adjusting a fader on a client app does not re-broadcast a FADER_CHANGE event (it does broadcast if a fader changes using Live)

- client Freeze event is not documented
 - 'FREEZE_ON'
 - 'FREEZE_OFF'
	
- Freeze event is not broadcast or re-broadcast.  (sending FREEZE_ON or FREEZE_OFF from a client app does not send a event to other client apps AND toggling FREEZE in Live does not send an event to client apps)

- client AutoBPM event is not documented
 - 'AUTO_BPM_ON'
 - 'AUTO_BPM_OFF'

- AutoBPM toggle event is not broadcast or re-broadcast.  (sending AUTO_BPM_ON or AUTO_BPM_OFF from a client app does not send an event to other client apps AND toggling AUTOBPM in Live does not send an event to client apps)


---


# Unknown Features

I need to test these on a PC (where AutoBPM works)

- Sending BEAT event does nothing? (windows only??)
- Have not seen a BPM or BEAT_OFF message (windows only??)


CHANGELOG.md
#Changelog

#0.2.0 - 2/20/2017
- Add example application link to README
- Switch faderChange event to fire with object containing index property instead of faderIndex
- Switch faderChange method to accept an object containing index property instead of faderName

#0.1.0 - 2/20/2017
- Fix faderChange event to fire with faderIndex and NOT faderName to match how Live sends the event
- Add example to scripts folder and to docs
- Fix example in README
- Add CHANGELOG

#0.0.1 - 2/20/2017
- Initial Release.
- Support for all known events defined in the Protocol Definition
- Support for all known methods defined in the Protocol Definition
- Support for undocumented methods: 'FREEZE_ON', 'FREEZE_OFF', 'AUTO_BPM_ON', 'AUTO_BPM_OFF' used in the Live Mobile App
- Mock server for finding undocumented events fired by the Live Mobile App
- 100% test coverage


test.js
const LightingController = require('./');

const myCtrl = new LightingController({
	password: '1234'
});

myCtrl.onAny((event, value) => {
	console.log(event, value);
});
myCtrl.on('connected', function() {
	console.log('connected')
	// myCtrl.cue('lasers');

	// setTimeout(function() {
	// 	myCtrl.cue('lasers')
	// },5000);
	myCtrl.buttonList();
	// 
	
	//myCtrl.bpm(400);
	//
});
// myCtrl.on('beatOn', function(){
// 	console.log('beatOn')
// 	console.log('feezing')
// 	setInterval(function() {
// 		console.log('sending beat')
// 		myCtrl.beat();
// 	}, 1000)
// });
myCtrl.on('buttonList', function(btns){
	console.log('buttonList')
	console.log(JSON.stringify(btns, null, 4));
	// myCtrl.freeze();
	// setTimeout(function() {
	// 	myCtrl.unfreeze();
	// },5000)
});



// //error
// myCtrl.on('error', function(err){
// 	console.log('error', err);
// });

// //unknown command
// myCtrl.on('unknownEvent', function(data) {
// 	console.log('unknown event', data)
// });

myCtrl.connect();



spec/helpers/index.js
const mockery = require('mockery');
mockery.enable({
	warnOnReplace: false,
    warnOnUnregistered: false
});

const netMock = {
	Socket: class Socket {
		constructor() {
			this.encoding = '';
			this.onMethods = {};
			this.writtenData = [];
		}

		setEncoding(encoding) {
			this.encoding = encoding;
		}

		on(name, cb) {
			this.onMethods[name] = cb;
		}
		connect(port, ip, callback) {
			this.connectCalled = arguments;
			callback();
		}
		write(data){
			this.writtenData.push(data);
		}
	}
};

mockery.registerMock('net', netMock);


spec/helpers/mock-server.js
// const xml = '<?xml version="1.0" encoding="UTF-8"?><buttons><page name="EFFECTS" columns="2" colbuttons_1="1" colbuttons_2="1"><button index="0" flash="0" pressed="0" line="1" column="1" color="#FFFFFF">lasers</button><button index="1" flash="0" pressed="0" line="1" column="2" color="#17E726">lightning</button></page><page name="ELLIPS" columns="1" colbuttons_1="6"><button index="2" flash="0" pressed="0" line="1" column="1" color="#FFFFFF">ellip_1</button><button index="3" flash="0" pressed="0" line="2" column="1" color="#FFFFFF">ellip_2</button><button index="4" flash="0" pressed="0" line="3" column="1" color="#FFFFFF">ellip_3</button><button index="5" flash="0" pressed="0" line="4" column="1" color="#FFFFFF">ellip_4</button><button index="6" flash="0" pressed="0" line="5" column="1" color="#FFFFFF">ellip_5</button><button index="7" flash="0" pressed="0" line="6" column="1" color="#FFFFFF">ellip_6</button></page><page name="GLOBAL" columns="2" colbuttons_1="1" colbuttons_2="2"><button index="8" flash="0" pressed="0" line="1" column="1" color="#000000">BLACKOUT</button><button index="9" flash="0" pressed="0" line="1" column="2" color="#2DD140">ALL~ELLIPS</button><button index="10" flash="0" pressed="0" line="2" column="2" color="#FFFFFF">WASH</button></page><page name="Page_7" columns="2" colbuttons_1="1" colbuttons_2="1"><button index="11" flash="0" pressed="0" line="1" column="1" color="#223CDC">baby_blue</button><button index="12" flash="0" pressed="0" line="1" column="2" color="#1035EE">demo</button></page><page name="SINGLE_LIGHT" columns="3" colbuttons_1="2" colbuttons_2="1" colbuttons_3="1"><button index="14" flash="0" pressed="0" line="2" column="1" color="#FFFFFF">FOH~BLUE</button><button index="15" flash="0" pressed="0" line="1" column="1" color="#FFFFFF">FOH~RED</button><button index="16" flash="0" pressed="0" line="1" column="2" color="#FFFFFF">R~PAR~BLUE</button><button index="17" flash="0" pressed="0" line="1" column="3" color="#FFFFFF">R~PAR~RED</button></page><page name="SPOTS" columns="1" colbuttons_1="1"><button index="18" flash="0" pressed="0" line="1" column="1" color="#FFFFFF">SPOT~OFF</button></page><page name="ZONES" columns="5" colbuttons_1="1" colbuttons_2="1" colbuttons_3="2" colbuttons_4="1" colbuttons_5="1"><button index="19" flash="0" pressed="0" line="1" column="1" color="#FFFFFF">LR TANK #1</button><button index="20" flash="0" pressed="0" line="1" column="3" color="#FFFFFF">LR CENTER #1</button><button index="21" flash="0" pressed="0" line="1" column="5" color="#FFFFFF">LR ELEVATOR #1</button><button index="22" flash="0" pressed="0" line="2" column="3" color="#FFFFFF">2/5</button><button index="23" flash="0" pressed="0" line="1" column="2" color="#FFFFFF">2/3</button><button index="24" flash="0" pressed="0" line="1" column="4" color="#FFFFFF">4/5</button></page><fader value="0">ELLIP DIMMER</fader><fader value="0">BACK PAR DIMMER</fader><fader value="0">Master fader 3</fader><fader value="0">Master fader 4</fader></buttons>';
// const CRLF = '\u000d\u000a';
// const SEPARATOR = '|';
// var net = require('net');

// var socket;

// var bufferedData = '';
// function onSocketData(data) {
// 	bufferedData += data;
// 	checkParse();
// }
// function checkParse() {
// 	const lineEnd = bufferedData.indexOf(CRLF);
// 	if (lineEnd > -1) {
// 		extractLine(lineEnd);
// 		//check again if we had a line to parse
// 		checkParse();
// 	}
// }

// function extractLine(lineEnd) {
// 	const line = bufferedData.substring(0, lineEnd);
// 	bufferedData = bufferedData.slice(lineEnd + CRLF.length);
// 	parseLine(line);
// }

// function parseLine(line) {
// 	let cmd = line;
// 	let data;
// 	if (line.indexOf(SEPARATOR) > -1) {
// 		const splitLine = line.split(SEPARATOR);
// 		cmd = splitLine[0];
// 		data = splitLine[1];
// 	}
// 	switch (cmd) {
// 		case 'HELLO':
// 			console.log('sent hello')
// 			socket.write('HELLO'+CRLF);
// 		break;
// 		case 'BUTTON_LIST':
// 			console.log('sending btns')
// 			socket.write('BUTTON_LIST'+SEPARATOR+ xml+CRLF);
// 		break;
// 		default:
// 		console.log(line);
// 		break;
// 	}
// }
// var server = net.createServer(function(s) {
// 	socket = s;
// 	s.on('data', onSocketData);
// });

// server.listen(7348, '192.168.0.4');


src/lighting-controller.js
const CRLF = '\u000d\u000a';
const SEPARATOR = '|';

var net = require('net');
const EventEmitter2 = require('eventemitter2').EventEmitter2;
const merge = require('utils-merge');
const xmlStringParser = require('xml2js').parseString;

const buttonListFormatter = require('./button-list-formatter');

/**
 * LightingController
 */
class LightingController extends EventEmitter2 {
	constructor(args) {
		const defaultSettings = {
			port: 7348,
			ip: '127.0.0.1',
			extApp: 'thelightingcontrollerclient',
			password: ''
		};
		super();
		this.settings = merge(defaultSettings, args);
		this.bufferedData = '';
		this.connected = false;
	}

	connect() {
		this.client = new net.Socket();
		this.client.setEncoding('utf8');
		this.client.on('data', data => this._onSocketData(data));
		this.client.on('close', () => this._onSocketClose());
		this.client.on('error', error => this._onSocketError(error));

		this.client.connect(this.settings.port, this.settings.ip, () => this._hello());
	}

	_onSocketData(data) {
		this.bufferedData += data;
		this._checkParse();
	}

	_onSocketClose() {
		this.connected = false;
		this.emit('disconnected');
	}

	_onSocketError(error) {
		this._onError({
			type: 'SOCKET',
			error: error
		});
	}

	_sendSocketMessage(message) {
		this.client.write(message + CRLF);
	}

	_checkParse() {
		const lineEnd = this.bufferedData.indexOf(CRLF);
		if (lineEnd > -1) {
			this._extractLine(lineEnd);
			//check again if we had a line to parse
			this._checkParse();
		}
	}

	_extractLine(lineEnd) {
		const line = this.bufferedData.substring(0, lineEnd);
		this.bufferedData = this.bufferedData.slice(lineEnd + CRLF.length);
		this._parseLine(line);
	}

	/**
	 * parseLine
	 * @param  {string} line the line to parse
	 *
	 * Network messages sent by Live application :
	 *	- "BPM" : Sent to the external application to ask the current BPM.
	 *	Response of the external application : "BPM|{bpm}".
	 
	 *	- "BEAT_ON" / “BEAT_OFF” : Sent to start/stop the sending of the real time beats by the
	 *	external application.
	 *	* Response of the external application : "BEAT" at each beat.
	 
	 *	- "BUTTON_PRESS|{button_name}" : Sent when a button is pressed in Live application.
	 *	(use if you need)
	 
	 *	- "BUTTON_RELEASE|{button_name}" : Sent when a button is released in Live
	 *	application.
	 *	(use if you need)
	 
	 *	- "FADER_CHANGE|{#fader}|" : Sent when a master fader position has changed in Live
	 *	application.
	 *	(use if you need)
	 
	 *	- “INTERFACE_CHANGE” : Sent to inform a change in Live interface (new button, button
	 *	delete, button rename, nb of master faders,, and so on …).
	 *	(use if you need)
	 */
	_parseLine(line) {
		let cmd = line;
		let data;
		if (line.indexOf(SEPARATOR) > -1) {
			const splitLine = line.split(SEPARATOR);
			cmd = splitLine[0];
			data = splitLine[1];
		}
		switch (cmd) {
			case 'HELLO':
				this.connected = true;
				this.emit('connected', data);
				break;
			case 'BEAT_ON':
				this.emit('beatOn', data);
				break;
			case 'BEAT_OFF':
				this.emit('beatOff', data);
				break;
			case 'BUTTON_LIST':
				this._onButtonList(data);
				break;
			case 'BUTTON_PRESS':
				this.emit('buttonPress', data);
				break;
			case 'BUTTON_RELEASE':
				this.emit('buttonRelease', data);
				break;
			case 'FADER_CHANGE':
				this.emit('faderChange', {
					index: Number(data),
					value: Number(line.split(SEPARATOR)[2])
				});
				break;
			case 'INTERFACE_CHANGE':
				this.emit('interfaceChange', data);
				break;
			case 'ERROR':
				this._onError({
					type: data
				});
				break;
			case 'BPM':
				this.emit('bpm', data);
				break;
			default:
				this.emit('unknownEvent', line);
				break;
		}
	}

	_onError(errorData) {
		if (!errorData.type) {
			errorData.type = 'UNKNOWN ERROR';
		}
		this.emit('error', errorData);
	}


	_onButtonList(line) {
		xmlStringParser(line, (err, result) => {
			if (err) {
				return this._onError({
					type: 'BUTTON LIST XML PARSE FAILED',
					error: err,
					data: line
				});
			}
			this.emit('buttonList', buttonListFormatter.format(result));
		});
	}

	/**
	 * @method hello
	 * 
	 * "HELLO|{ExtApp}|{password}" : Must be sent immediately after the TCP connection
	 * has established otherwise the response "ERROR|NOT LOGGED" will be returned.
	 * {ExtApp} is the name of the external application. The password has to be sent in clear.
	 * Response of Live : "HELLO" if OK else "ERROR|BAD PASSWORD".
	 * 
	 */
	_hello() {
		this._sendSocketMessage('HELLO|' + this.settings.extApp + SEPARATOR + this.settings.password);
	}

	/**
	 * @method bpm
	 *
	 * "BPM|{bpm}" : Sent at each song start or at each "BPM" message from Live.
	 * This assigns the new BPM value in the “manual BPM” section of Live.
	 */
	bpm(bpm) {
		const num = Number(bpm);
		if (isNaN(num) || num < 0) {
			return this._onError({
				type: 'CLIENT ERROR',
				error: 'invalid bpm value'
			});
		}
		this._sendSocketMessage('BPM' + SEPARATOR + bpm);
	}

	/**
	 * @method beat
	 * 
	 * "BEAT" : Sent at each beat after receiving a "BEAT_ON" message from Live and stopped
	 * after receiving a "BEAT_OFF" message from Live.
	 * This increments the beat in the “auto BPM” section of Live.
	 */
	beat() {
		this._sendSocketMessage('BEAT');
	}

	freeze() {
		this._sendSocketMessage('FREEZE_ON');
	}

	unfreeze() {
		this._sendSocketMessage('FREEZE_OFF');
	}

	autoBpmOn() {
		this._sendSocketMessage('AUTO_BPM_ON');
	}

	autoBpmOff() {
		this._sendSocketMessage('AUTO_BPM_OFF');
	}

	/**
	 * @method cue
	 *
	 * "CUE|{cue_name}" : Sent at each meeting of a cue point in the external application.
	 * This toggles the button with the name {cue_name} in Live.
	 *
	 * note: This just toggles whatever button name is passed to it.  TLC software will
	 * still fire buttonPress/buttonRelease events to the client when this is used.  Also
	 * aliased as the (more accurate) buttonToggle.  This name is just to match the official docs.
	 * 
	 */
	cue(cueName) {
		if (!cueName) {
			return this._onError({
				type: 'CLIENT ERROR',
				error: 'invalid cue name'
			});
		}
		this._sendSocketMessage('CUE' + SEPARATOR + cueName);
	}

	/**
	 * @method buttonToggle
	 * @aliasOf cue
	 */
	buttonToggle(buttonName) {
		this.cue(buttonName);
	}

	/**
	 * @method buttonPress
	 *
	 * "BUTTON_PRESS|{button_name}" : Sent when a button is pressed in the external application.
	 * This presses the button with the name {button_name} in Live.
	 */
	buttonPress(buttonName) {
		if (!buttonName) {
			return this._onError({
				type: 'CLIENT ERROR',
				error: 'invalid button name'
			});
		}
		this._sendSocketMessage('BUTTON_PRESS' + SEPARATOR + buttonName);
	}

	/**
	 * @method buttonRelease
	 *
	 * "BUTTON_RELEASE|{button_name}" : Sent when a button is released in the external application.
	 * This releases the button with the name {button_name} in Live.
	 */
	buttonRelease(buttonName) {
		if (!buttonName) {
			return this._onError({
				type: 'CLIENT ERROR',
				error: 'invalid button name'
			});
		}
		this._sendSocketMessage('BUTTON_RELEASE' + SEPARATOR + buttonName);
	}

	/**
	 * @method faderChange
	 *
	 * "FADER_CHANGE|{#fader}|{position}" : Sent when a master fader position has changed in the
	 * external application.
	 * This changes the position of the master fader in Live
	 */
	faderChange(faderIndex, position) {
		faderIndex = Number(faderIndex);
		if (isNaN(faderIndex)) {
			return this._onError({
				type: 'CLIENT ERROR',
				error: 'invalid fader index'
			});
		}
		position = Number(position);
		if (isNaN(position) || position > 100 || position < -100) {
			return this._onError({
				type: 'CLIENT ERROR',
				error: 'invalid fader value'
			});
		}
		this._sendSocketMessage('FADER_CHANGE' + SEPARATOR + faderIndex + SEPARATOR + position);
	}

	/**
	 * @method buttonList
	 *
	 * "BUTTON_LIST" : Sent to ask Live for its button and master faders list
	 * Response of Live : “BUTTON_LIST|” followed by a XML feed with the following format :
	 *
	 *	 <?xml version="1.0" encoding="UTF-8"?>
	 *		<buttons>
	 *			<page name="head" columns="2" colbuttons_1="2" colbuttons_2="2">
	 *				<button index="0" flash="0" pressed="0" line="1" column="1" color="#80FFFF">
	 *					Button name 1
	 *				</button>
	 *				<button index="1" flash="0" pressed="0" line="2" column="1" color="#80FFFF">
	 *					Button name 2
	 *				</button>
	 *				.
	 *				.
	 *				.
	 *			</page>
	 *			.
	 *			.
	 *			.
	 *			<fader value=”0”>
	 *				Fader name 1
	 *			</fader>
	 *			<fader value=”50”>
	 *				Fader name 2
	 *			</fader>
	 *			.
	 *			.
	 *			.
	 *
	 *	<page> balise parameters :
	 *	“name” : name of the page
	 *	“columns” : nb columns in page in Live application
	 *	“colbuttons_x” : nb of buttons in the column x in Live application
	 *		
	 *	<button> balise parameters :
	 *	“index” : numeric id of the button
	 *	“flash” : flash type button (monostable – push button)
	 *	“pressed” : state of the button : 0 = released, 1 = pressed
	 *	“line” : line of the button in Live application
	 *	“column” : column of the button in Live application
	 *	“color” : color of the button in Live application (HTML format)
	 *
	 *	<fader> balise parameters :
	 *	“value” : numeric value of the fader position
	 *	(use only parameters you need)
	 */
	buttonList() {
		this._sendSocketMessage('BUTTON_LIST');
	}
}

module.exports = LightingController;


src/button-list-formatter.js
function formatButtonListObject(rawButtonList) {
    const returnObject = {
        pages: [],
        faders: []
    };
    
    if (rawButtonList.buttons && rawButtonList.buttons.page) {
        rawButtonList.buttons.page.forEach((page) => {
            if (!page.$) { return; }
            const pageObject = {
                name: page.$.name,
                columns: page.$.columns,
                columnButtons: {},
                buttons: []
            }
            var col = 1;
            while (col <= pageObject.columns) {
                if (page.$['colbuttons_' + col]) {
                    pageObject.columnButtons[col] = page.$['colbuttons_' + col];
                }
                col++;
            }
            if (page.button && page.button.length) {
                page.button.forEach((button) => {
                    if (!button.$) { return; }
                    pageObject.buttons.push({
                        name: button['_'] || '',
                        index: Number(button.$.index),
                        flash: Boolean(Number(button.$.flash)),
                        pressed: Boolean(Number(button.$.pressed)),
                        line: Number(button.$.line),
                        column: Number(button.$.column),
                        color: button.$.color
                    });
                });
            }
            returnObject.pages.push(pageObject);
        });
    }
    if (rawButtonList.buttons && rawButtonList.buttons.fader) {
        rawButtonList.buttons.fader.forEach((fader) => {
            if (!fader.$) { return; }
            returnObject.faders.push({
                name: fader['_'],
                value: Number(fader.$.value)
            });
        });
    }
    return returnObject;
}

module.exports = {
    format: formatButtonListObject
};


scripts/mock-server.js
const xml = '<?xml version="1.0" encoding="UTF-8"?><buttons><page name="EFFECTS" columns="2" colbuttons_1="1" colbuttons_2="1"><button index="0" flash="0" pressed="0" line="1" column="1" color="#FFFFFF">lasers</button><button index="1" flash="0" pressed="0" line="1" column="2" color="#17E726">lightning</button></page><page name="ELLIPS" columns="1" colbuttons_1="6"><button index="2" flash="0" pressed="0" line="1" column="1" color="#FFFFFF">ellip_1</button><button index="3" flash="0" pressed="0" line="2" column="1" color="#FFFFFF">ellip_2</button><button index="4" flash="0" pressed="0" line="3" column="1" color="#FFFFFF">ellip_3</button><button index="5" flash="0" pressed="0" line="4" column="1" color="#FFFFFF">ellip_4</button><button index="6" flash="0" pressed="0" line="5" column="1" color="#FFFFFF">ellip_5</button><button index="7" flash="0" pressed="0" line="6" column="1" color="#FFFFFF">ellip_6</button></page><page name="GLOBAL" columns="2" colbuttons_1="1" colbuttons_2="2"><button index="8" flash="0" pressed="0" line="1" column="1" color="#000000">BLACKOUT</button><button index="9" flash="0" pressed="0" line="1" column="2" color="#2DD140">ALL~ELLIPS</button><button index="10" flash="0" pressed="0" line="2" column="2" color="#FFFFFF">WASH</button></page><page name="Page_7" columns="2" colbuttons_1="1" colbuttons_2="1"><button index="11" flash="0" pressed="0" line="1" column="1" color="#223CDC">baby_blue</button><button index="12" flash="0" pressed="0" line="1" column="2" color="#1035EE">demo</button></page><page name="SINGLE_LIGHT" columns="3" colbuttons_1="2" colbuttons_2="1" colbuttons_3="1"><button index="14" flash="0" pressed="0" line="2" column="1" color="#FFFFFF">FOH~BLUE</button><button index="15" flash="0" pressed="0" line="1" column="1" color="#FFFFFF">FOH~RED</button><button index="16" flash="0" pressed="0" line="1" column="2" color="#FFFFFF">R~PAR~BLUE</button><button index="17" flash="0" pressed="0" line="1" column="3" color="#FFFFFF">R~PAR~RED</button></page><page name="SPOTS" columns="1" colbuttons_1="1"><button index="18" flash="0" pressed="0" line="1" column="1" color="#FFFFFF">SPOT~OFF</button></page><page name="ZONES" columns="5" colbuttons_1="1" colbuttons_2="1" colbuttons_3="2" colbuttons_4="1" colbuttons_5="1"><button index="19" flash="0" pressed="0" line="1" column="1" color="#FFFFFF">LR TANK #1</button><button index="20" flash="0" pressed="0" line="1" column="3" color="#FFFFFF">LR CENTER #1</button><button index="21" flash="0" pressed="0" line="1" column="5" color="#FFFFFF">LR ELEVATOR #1</button><button index="22" flash="0" pressed="0" line="2" column="3" color="#FFFFFF">2/5</button><button index="23" flash="0" pressed="0" line="1" column="2" color="#FFFFFF">2/3</button><button index="24" flash="0" pressed="0" line="1" column="4" color="#FFFFFF">4/5</button></page><fader value="0">ELLIP DIMMER</fader><fader value="0">BACK PAR DIMMER</fader><fader value="0">Master fader 3</fader><fader value="0">Master fader 4</fader></buttons>';
const CRLF = '\u000d\u000a';
const SEPARATOR = '|';
var net = require('net');

var socket;

var bufferedData = '';
function onSocketData(data) {
	bufferedData += data;
	checkParse();
}
function checkParse() {
	const lineEnd = bufferedData.indexOf(CRLF);
	if (lineEnd > -1) {
		extractLine(lineEnd);
		//check again if we had a line to parse
		checkParse();
	}
}

function extractLine(lineEnd) {
	const line = bufferedData.substring(0, lineEnd);
	bufferedData = bufferedData.slice(lineEnd + CRLF.length);
	parseLine(line);
}

function parseLine(line) {
	let cmd = line;
	let data;
	if (line.indexOf(SEPARATOR) > -1) {
		const splitLine = line.split(SEPARATOR);
		cmd = splitLine[0];
		data = splitLine[1];
	}
	switch (cmd) {
		case 'HELLO':
			console.log('sent hello')
			socket.write('HELLO'+CRLF);
		break;
		case 'BUTTON_LIST':
			console.log('sending btns')
			socket.write('BUTTON_LIST'+SEPARATOR+ xml+CRLF);
		break;
		default:
		console.log(line);
		break;
	}
}
var server = net.createServer(function(s) {
	socket = s;
	s.on('data', onSocketData);
});

server.listen(7348, '192.168.0.4');
